### ISSUE TITLE
Rewrite `expand_pointer_allow_strip` and `expand_macro_children`

### ISSUE THREAD

**Issue by CohenArthur** [Rust-GCC/gccrs#1036]
is there a way we could rewrite this for loop so it always continues to the  end even if we miss incremenentations in the body?

If you hit a case in the fort loop where you would want the iteratator not to move forward you could always issue it--;

_Originally posted by @philberty in https://github.com/Rust-GCC/gccrs/pull/1029#discussion_r828964817_

**Comment by tlietz** [comment]
I am here from Zulip.

**Comment by tlietz** [comment]
I think that the current implementation might be the best way to do this because the combination of incrementing in the for loop header, and `.erase()` can lead to undefined behavior.

For example, suppose we tried refactoring the `for` loop in the `expand_pointer_allow_strip` function to the following:
```cpp
for (auto it = values.begin (); it != values.end (); ++it)
  {
     auto &value = *it;

     // mark for stripping if required
     value->accept_vis (*this);
     if (value->is_marked_for_strip ())
     {
	it = values.erase (it);
        --it;
    }
  }
```

If the first iterator of the `values` vector is erased, `--it` would be undefined. 

I may be missing something though and would love feedback on this.

**Comment by CohenArthur** [comment]
Right! Well thank you for trying to resolve this. I guess we could add more check to make sure we're not decrementing the iterator on `values.begin()`, but it might get even messier than the current implementation. If you can think about another solution to clean this up, feel free to ask for feedback too!
